billing/actions/accounts.py

```diff
@@ -220,3 +228,123 @@ def assign_funds_to_invoice(invoice_id: str) -> bool:
+    :param account_ids: List of account ids to be evaluated
+    :param unpaid_invoices_threshold: Number of unpaid invoices to consider an user as
+                                      a delinquent
+    :param days_since_last_unpaid_threshold: Days to take into account since the last
+                                             unpaid invoice to consider an user as a
+                                             delinquent
+    :param currency_amount_threshold_map: Balance threshold to consider an user as a
+                                          delinquent.
+                                          Ex: {'CHF': 200, 'EUR': 100, 'NOK': 150}
+    :return: New delinquent accounts and reason Ex: {111: 'Reason why', ...}
+    """
+    legal_accounts = Account.objects.filter(id__in=account_ids, delinquent=False)
+    new_delinquent_accounts_map = {}
+    for account in legal_accounts:
+        is_delinquent, reason = is_a_delinquent_account(
+            account,
+            unpaid_invoices_threshold,
+            days_since_last_unpaid_threshold,
+            currency_amount_threshold_map,
+        )
+        if is_delinquent:
+            new_delinquent_accounts_map[account.id] = reason
+
+    Account.objects.filter(
+        id__in=new_delinquent_accounts_map.keys()
+    ).update(delinquent=True)
+    return new_delinquent_accounts_map
+
+
+def mark_accounts_as_legal(
+    account_ids: List[int],
+    unpaid_invoices_threshold: Optional[int],
+    days_since_last_unpaid_threshold: Optional[int],
+    currency_amount_threshold_map: Optional[dict],
+) -> List[int]:
+    """
+    Mark accounts as delinquent when some criteria are accomplished
+    :param account_ids: List of account ids to be evaluated
+    :param unpaid_invoices_threshold: Number of unpaid invoices to consider an user as
+                                      a delinquent
+    :param days_since_last_unpaid_threshold: Days to take into account since the last
+                                             unpaid invoice to consider an user as a
+                                             delinquent
+    :param currency_amount_threshold_map: Balance threshold to consider an user as a
+                                          delinquent.
+                                          Ex: {'CHF': 200, 'EUR': 100, 'NOK': 150}
+    :return: Legalized account ids
+    """
+    delinquent_accounts = Account.objects.filter(id__in=account_ids, delinquent=True)
+    legalized_accounts_ids = []
+    for account in delinquent_accounts:
+        is_delinquent, _ = is_a_delinquent_account(
+            account,
+            unpaid_invoices_threshold,
+            days_since_last_unpaid_threshold,
+            currency_amount_threshold_map,
+        )
+        if not is_delinquent:
+            legalized_accounts_ids.append(account.id)
+
+    Account.objects.filter(id__in=legalized_accounts_ids).update(delinquent=False)
+    return legalized_accounts_ids
+
+
+def is_a_delinquent_account(
```

Comment:
Not sure about this but instead of short cutting and returning True as soon as one of the condition for delinquency is met, you could return a list of delinquency reasons (and in the case where the account is not delinquent the list would be empty)?

Reply:
Yes, probably better option